File: tools/AutoTest/AutoTest/AutoTestScript/AutoTestBlockTransfer.lua
function assertResult( self,  aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
	print("lua aBOOL=" .. tostring(aBOOL) .. " aInt=" .. tostring(aInt) .. " aInteger=" .. tostring(aInteger) .. " aFloat=" .. tostring(aFloat) .. " aCGFloat=" .. tostring(aCGFloat) .. " aId=" .. tostring(aId))
	assert(aBOOL==TEST_VALUE_BOOL);assert(aInt==TEST_VALUE_INT);
	assert(aInteger==TEST_VALUE_INTEGER);assert(isDoubleEqual(aFloat, TEST_VALUE_FLOAT));
	assert(isDoubleEqual(aCGFloat, TEST_VALUE_CGFLOAT));
	assert(aId == self);
end


waxClass{"AutoTestBlockTransfer"}

function autoTestStart( self )
	self:testReturnVoidWithVoidBlock(
	toblock(
		function()
			print("LUA TEST SUCCESS: testReturnVoidWithVoidBlock")
		end
		, {"void"})
	)

	self:testReturnVoidWithFirstIntBlock(
	toblock(
		function(aFirstInt, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("aFirstInt=" .. tostring(aFirstInt))
			assert(aFirstInt == TEST_VALUE_INT, "aFirstInt不等")
			assertResult(self, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("LUA TEST SUCCESS: testReturnVoidWithFirstIntBlock")
		end
		, {"void","int", "BOOL", "int", "NSInteger", "float", "CGFloat", "id"})
	)
	self:testReturnBOOLWithFirstBOOLBlock(
	toblock(
		function(aFirstBOOL, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("aFirstBOOL=" .. tostring(aFirstBOOL))
			assert(aFirstBOOL == TEST_VALUE_BOOL)
			assertResult(self, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("LUA TEST SUCCESS: testReturnBOOLWithFirstBOOLBlock")
			return aFirstBOOL
		end
		, {"BOOL","BOOL", "BOOL", "int", "NSInteger", "float", "CGFloat", "id"})
	)
	self:testReturnIntWithFirstIntBlock(
	toblock(
		function(aFirstInt, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("aFirstInt=" .. tostring(aFirstInt))
			assert(aFirstInt == TEST_VALUE_INT, "aFirstInt不等")
			assertResult(self, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("LUA TEST SUCCESS: testReturnIntWithFirstIntBlock")
			return aFirstInt
		end
		, {"int","int", "BOOL", "int", "NSInteger", "float", "CGFloat", "id"})
	)

	self:testReturnIntegerWithFirstIntegerBlock(
	toblock(
		function(aFirstInteger, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("aFirstInteger=" .. tostring(aFirstInteger))
			assert(aFirstInteger == TEST_VALUE_INTEGER)
			assertResult(self, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("LUA TEST SUCCESS: testReturnIntegerWithFirstIntegerBlock")
			return aFirstInteger
		end
		, {"NSInteger","NSInteger", "BOOL", "int", "NSInteger", "float", "CGFloat", "id"})
	)

	self:testReturnCharPointerWithFirstCharPointerBlock(
	toblock(
		function(aFirstCharPointer, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("aFirstCharPointer=" .. tostring(aFirstPointer))
			assertResult(self, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("LUA TEST SUCCESS: testReturnCharPointerWithFirstCharPointerBlock")
			return aFirstCharPointer
		end
		, {"char *","char*", "BOOL", "int", "NSInteger", "float", "CGFloat", "id"})
	)

	self:testReturnVoidPointerWithFirstVoidPointerBlock(
	toblock(
		function(aFirstVoidPointer, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("aFirstVoidPointer=" .. tostring(aFirstVoidPointer))
			-- assert(aFirstVoidPointer == TEST_VALUE_INT, "aFirstInt不等")
			assertResult(self, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("LUA TEST SUCCESS: testReturnVoidPointerWithFirstVoidPointerBlock")
			return aFirstVoidPointer
		end
		, {"void *","void*", "BOOL", "int", "NSInteger", "float", "CGFloat", "id"})
	)

	self:testReturnIdWithFirstIdBlock(
	toblock(
		function(aFirstId, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("aFirstId=" .. tostring(aFirstId))
			-- assert(aFirstId == self, "aFirstInt不等")
			assertResult(self, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("LUA TEST SUCCESS: testReturnIdWithFirstIdBlock")
			return aFirstId
		end
		, {"id","id", "BOOL", "int", "NSInteger", "float", "CGFloat", "id"})
	)



	self:testReturnFloatWithFirstFloatBlock(
	toobjc(
		function(aFirstFloat, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("aFirstFloat=" .. tostring(aFirstFloat))
			assert(isDoubleEqual(aFirstFloat, TEST_VALUE_FLOAT));
			assertResult(self, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("LUA TEST SUCCESS: testReturnFloatWithFirstFloatBlock")
			return aFirstFloat
		end
		):luaBlockWithParamsTypeArray({"float","float", "BOOL", "int", "NSInteger", "float", "CGFloat", "id"})
	)
	self:testReturnCGFloatWithFirstCGFloatBlock(
	toobjc(
		function(aFirstCGFloat, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("aFirstCGFloat=" .. tostring(aFirstCGFloat))
			assert(isDoubleEqual(aFirstCGFloat, TEST_VALUE_CGFLOAT));
			assertResult(self, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("LUA TEST SUCCESS: testReturnCGFloatWithFirstCGFloatBlock")
			return aFirstCGFloat
		end
		):luaBlockWithParamsTypeArray({"CGFloat","CGFloat", "BOOL", "int", "NSInteger", "float", "CGFloat", "id"})
	)
	self:testReturnDoubleWithFirstDoubleBlock(
	toobjc(
		function(aFirstDouble, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("aFirstDouble=" .. tostring(aFirstDouble))
			assert(isDoubleEqual(aFirstDouble, TEST_VALUE_DOUBLE));
			assertResult(self, aBOOL, aInt, aInteger, aFloat, aCGFloat, aId)
			print("LUA TEST SUCCESS: testReturnDoubleWithFirstDoubleBlock")
			return aFirstDouble
		end
		):luaBlockWithParamsTypeArray({"double","double", "BOOL", "int", "NSInteger", "float", "CGFloat", "id"})
	)
end

File: tools/AutoTest/AutoTest/AutoTestClass/AutoTestBlockTransfer.m
//
//  AutoTestBlockTransfer.m
//  
//
//  Created by junzhan on 15-1-6.
//  Copyright (c) 2015年 junzhan. All rights reserved.
//

#import "AutoTestBlockTransfer.h"
#import "AutoTestUtil.h"
@implementation AutoTestBlockTransfer

- (void)autoTestStart{
//lua里翻译这些block逻辑
    [self testReturnVoidWithFirstIntBlock:^(int aFirstInt, BOOL aBOOL, int aInt, NSInteger aInteger, float aFloat, CGFloat aCGFloat, id aId) {
        NSLog(@"ok");
    }];
}

- (void)testReturnVoidWithVoidBlock:(void(^)())block{
    block();
    NSLog(@"OC TEST SUCCESS: %s", __PRETTY_FUNCTION__);
}



//i^csiqfdB@@
- (void)testReturnVoidWithFirstIntBlock:(void(^)(int aFirstInt, BOOL aBOOL, int aInt, NSInteger aInteger, float aFloat, CGFloat aCGFloat, id aId))block{

    block(TEST_VALUE_INT, TEST_VALUE_BOOL, TEST_VALUE_INT, TEST_VALUE_INTEGER, TEST_VALUE_FLOAT,TEST_VALUE_CGFLOAT, self);
    NSLog(@"OC TEST SUCCESS: %s", __PRETTY_FUNCTION__);
}

//B^csiqfdB@@
- (void)testReturnBOOLWithFirstBOOLBlock:(BOOL(^)(BOOL aFirstBOOL,  BOOL aBOOL, int aInt, NSInteger aInteger, float aFloat, CGFloat aCGFloat, id aId))block{
    BOOL res = block(TEST_VALUE_BOOL, TEST_VALUE_BOOL, TEST_VALUE_INT, TEST_VALUE_INTEGER, TEST_VALUE_FLOAT,TEST_VALUE_CGFLOAT, self);
    NSLog(@"res=%d F=%s LINE=%d", res, __PRETTY_FUNCTION__, __LINE__);
    NSAssert(res == TEST_VALUE_BOOL, @"%s返回值错误", __PRETTY_FUNCTION__);
    NSLog(@"OC TEST SUCCESS: %s", __PRETTY_FUNCTION__);
}

- (void)testReturnIntWithFirstIntBlock:(int(^)(int aFirstInt, BOOL aBOOL, int aInt, NSInteger aInteger, float aFloat, CGFloat aCGFloat, id aId))block{
    int res = block(TEST_VALUE_INT, TEST_VALUE_BOOL, TEST_VALUE_INT, TEST_VALUE_INTEGER, TEST_VALUE_FLOAT,TEST_VALUE_CGFLOAT, self);
    NSLog(@"res=%d F=%s LINE=%d", res, __PRETTY_FUNCTION__, __LINE__);
    NSAssert(res == TEST_VALUE_INT, @"返回值错误");
    NSLog(@"OC TEST SUCCESS: %s", __PRETTY_FUNCTION__);
}

- (void)testReturnIntegerWithFirstIntegerBlock:(NSInteger(^)(NSInteger aFirstInteger, BOOL aBOOL, int aInt, NSInteger aInteger, float aFloat, CGFloat aCGFloat, id aId))block{
    NSInteger res = block(TEST_VALUE_INTEGER, TEST_VALUE_BOOL, TEST_VALUE_INT, TEST_VALUE_INTEGER, TEST_VALUE_FLOAT,TEST_VALUE_CGFLOAT, self);
    NSLog(@"res=%ld F=%s LINE=%d", res, __PRETTY_FUNCTION__, __LINE__);
    NSAssert(res == TEST_VALUE_INTEGER, @"返回值错误");
    NSLog(@"OC TEST SUCCESS: %s", __PRETTY_FUNCTION__);
}

- (void)testReturnCharPointerWithFirstCharPointerBlock:(char *(^)(char *aFirstCharPointer, BOOL aBOOL, int aInt, NSInteger aInteger, float aFloat, CGFloat aCGFloat, id aId))block{
    char *res = block(TEST_VALUE_CHAR_POINTER, TEST_VALUE_BOOL, TEST_VALUE_INT, TEST_VALUE_INTEGER, TEST_VALUE_FLOAT,TEST_VALUE_CGFLOAT, self);
    NSLog(@"res=%s F=%s LINE=%d", res, __PRETTY_FUNCTION__, __LINE__);
    NSAssert(strcmp(res, TEST_VALUE_CHAR_POINTER) == 0, @"%s返回值错误", __PRETTY_FUNCTION__);
    NSLog(@"OC TEST SUCCESS: %s", __PRETTY_FUNCTION__);
}

- (void)testReturnVoidPointerWithFirstVoidPointerBlock:(void *(^)(void *aFirstVoidPointer, BOOL aBOOL, int aInt, NSInteger aInteger, float aFloat, CGFloat aCGFloat, id aId))block{
    void *res = block(TEST_VALUE_VOID_POINTER, TEST_VALUE_BOOL, TEST_VALUE_INT, TEST_VALUE_INTEGER, TEST_VALUE_FLOAT,TEST_VALUE_CGFLOAT, self);
    NSLog(@"res=%p F=%s LINE=%d", res, __PRETTY_FUNCTION__, __LINE__);
    NSAssert(res == TEST_VALUE_VOID_POINTER, @"%s返回值错误", __PRETTY_FUNCTION__);
    NSLog(@"OC TEST SUCCESS: %s", __PRETTY_FUNCTION__);
}

- (void)testReturnIdWithFirstIdBlock:(id(^)(id aFirstId, BOOL aBOOL, int aInt, NSInteger aInteger, float aFloat, CGFloat aCGFloat, id aId))block{
    UIViewController *vc = [[UIViewController alloc] init];
    id res = block(vc, TEST_VALUE_BOOL, TEST_VALUE_INT, TEST_VALUE_INTEGER, TEST_VALUE_FLOAT,TEST_VALUE_CGFLOAT, self);
    NSLog(@"res=%@ F=%s LINE=%d", res, __PRETTY_FUNCTION__, __LINE__);
    NSAssert([res isKindOfClass:[UIViewController class]], @"%s返回值错误", __PRETTY_FUNCTION__);
    NSLog(@"OC TEST SUCCESS: %s", __PRETTY_FUNCTION__);
}

#pragma mark float

- (void)testReturnFloatWithFirstFloatBlock:(float(^)(float aFirstFloat, BOOL aBOOL, int aInt, NSInteger aInteger, float aFloat, CGFloat aCGFloat, id aId))block{
    float res = block(TEST_VALUE_FLOAT, TEST_VALUE_BOOL, TEST_VALUE_INT, TEST_VALUE_INTEGER, TEST_VALUE_FLOAT,TEST_VALUE_CGFLOAT, self);
    NSLog(@"res=%f F=%s LINE=%d", res, __PRETTY_FUNCTION__, __LINE__);
    NSAssert([AutoTestUtil isDoubleEqual:res aDouble:TEST_VALUE_FLOAT], @"%s返回值错误", __PRETTY_FUNCTION__);
    NSLog(@"OC TEST SUCCESS: %s", __PRETTY_FUNCTION__);
}

- (void)testReturnCGFloatWithFirstCGFloatBlock:(CGFloat(^)(CGFloat aFirstCGFloat, BOOL aBOOL, int aInt, NSInteger aInteger, float aFloat, CGFloat aCGFloat, id aId))block{
    CGFloat res = block(TEST_VALUE_CGFLOAT, TEST_VALUE_BOOL, TEST_VALUE_INT, TEST_VALUE_INTEGER, TEST_VALUE_FLOAT,TEST_VALUE_CGFLOAT, self);
    NSLog(@"res=%f F=%s LINE=%d", res, __PRETTY_FUNCTION__, __LINE__);
    NSAssert([AutoTestUtil isDoubleEqual:res aDouble:TEST_VALUE_CGFLOAT], @"%s返回值错误", __PRETTY_FUNCTION__);
    NSLog(@"OC TEST SUCCESS: %s", __PRETTY_FUNCTION__);
}

- (void)testReturnDoubleWithFirstDoubleBlock:(double(^)(double aFirstDouble, BOOL aBOOL, int aInt, NSInteger aInteger, float aFloat, CGFloat aCGFloat, id aId))block{
    
    double res = block(TEST_VALUE_DOUBLE, TEST_VALUE_BOOL, TEST_VALUE_INT, TEST_VALUE_INTEGER, TEST_VALUE_FLOAT,TEST_VALUE_CGFLOAT, self);
    NSLog(@"res=%f F=%s LINE=%d", res, __PRETTY_FUNCTION__, __LINE__);
    NSAssert([AutoTestUtil isDoubleEqual:res aDouble:TEST_VALUE_DOUBLE], @"%s返回值错误", __PRETTY_FUNCTION__);
    NSLog(@"OC TEST SUCCESS: %s", __PRETTY_FUNCTION__);
}
@end
